Add template to evaluate clustering to Ewing cell typing

allyhawkins opened the pull request:
### Purpose/implementation Section
#### Please link to the GitHub issue that this pull request addresses.

Closes #685 

#### What is the goal of this pull request?

We want to be able to assign clusters to each of the libraries in SCPCP000015. There is no one size fits all with clustering, so here I am adding a template notebook that can be used to perform clustering across a set of parameters and plot the results. Ultimately, we can render this notebook for all libraries in the project and use the rendered report to choose the best clusters for that library. 


#### Briefly describe the general approach you took to achieve this goal.

This is mostly based on the first section of the exploratory notebook added in #683, but I turned it into a template that looks at only a single library at a time. For each library, Louvain, Jaccard clusters are calculated across a range of K. Then we look at silhouette width, purity, and stability for each parameter and plot the results. 

I also included a section where we compare the clusters to the SingleR annotations obtained from `aucell-singler-annotation.sh`. Because this notebook will be somewhat dependent on running that workflow I actually think I might add rendering this as the last step in that workflow rather than creating a whole separate workflow? I didn't do that in this PR, but that is why I am adding this as `03-clustering.Rmd` to the `template_notebooks/aucell-singler-workflow` folder. Let me know if you think I should move this and ultimately create a separate workflow. 

The last section looks at marker gene expression for the different cell classes that we have marker genes for. I chose to print these density plots for every value of K, because I think in some cases where we see multiple values of K have similar stats, it could be helpful to at least ensure we have clear tumor/normal separation. I would expect good clustering to separate the tumor cells from the normal cells. 

Note that I also addressed the spelling errors for Ewing's and metacells from #691. Let me know if you disagree with any choices made. 

#### If known, do you anticipate filing additional pull requests to complete this analysis module?

Yes 

### Results

No results here, but I rendered this report with one of the libraries and am including it here to help with review: 
[03-clustering.html.zip](https://github.com/user-attachments/files/16502127/03-clustering.html.zip)


### Provide directions for reviewers
#### What are the software and computational requirements needed to be able to run the code in this PR?

This just runs locally and uses the renv environment for this module. 



### Author checklists

_Check all those that apply._
_Note that you may find it easier to check off these items after the pull request is actually filed._


#### Analysis module and review

- [ ] This analysis module [uses the analysis template and has the expected directory structure](https://openscpca.readthedocs.io/en/latest/contributing-to-analyses/analysis-modules/).
- [ ] The analysis module `README.md` has been updated to reflect code changes in this pull request.
- [x] The analytical code is documented and contains comments.
- [ ] Any results and/or plots this code produces have been added to your S3 bucket for review.

#### Reproducibility checklist

- [ ] Code in this pull request has been added to the GitHub Action workflow that runs this module.
- [ ] The dependencies required to run the code in this pull request have been added to the analysis module `Dockerfile`.
- [ ] If applicable, the dependencies required to run the code in this pull request have been added to the analysis module conda `environment.yml` file.
- [ ] If applicable, R package dependencies required to run the code in this pull request have been added to the analysis module `renv.lock` file.

sjspielman commented on analyses/cell-type-ewings/exploratory_analysis/05-cluster-exploration.Rmd:
I think it's worth removing the function definitions from this file and sourcing the new `utils/clustering-functions.R` instead

allyhawkins replied:
The functions in this notebook are a tiny bit different because they take into account the fact that we are looking at multiple libraries in some cases. So I didn't want to source them in here and have to make a bunch of modifications to this notebook.

sjspielman commented on analyses/cell-type-ewings/scripts/README.md:
```diff
@@ -239,5 +239,9 @@ The `utils` folder contains scripts with any helper functions that are used in `
+2. `tumor-validation-helpers.R`: These functions are used in the template notebooks in `template_notebooks/auc-singler-workflow` and the notebooks found in the `exploratory_analysis/annotation_notebooks` folder.
    The functions are helpful for exploring and validate tumor cell annotations in individual samples.
+   
+3. `clustering-functions.R`: These functions are used in the template notebook for exploring clustering, `template_notebooks/auc-singler-workflow/03-clustering.Rmd`. 
```
If you remove functions from the exploratory notebook, also indicate here that the util script is used there too.

sjspielman commented on analyses/cell-type-ewings/scripts/utils/clustering-functions.R:
```diff
@@ -0,0 +1,212 @@
+  return(clusters)
+}
+
+# define a function to perform clustersweep and get clusters across multiple values of k
```
I'd add `(5,40,5)` into this comment since it's hard-coded (which I don't see as a problem necessarily)

sjspielman commented on analyses/cell-type-ewings/template_notebooks/auc-singler-workflow/03-clustering.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+
+Below we visualize the cluster assignments using each parameter on a UMAP. 
+This can be helpful to see any values of K that may have obvious over or under clustering. 
```
```suggestion
This can be helpful to see any values of `k` that may have obvious over or under clustering. 
```

sjspielman commented on analyses/cell-type-ewings/template_notebooks/auc-singler-workflow/03-clustering.Rmd:
```diff
@@ -0,0 +1,265 @@
+  theme(
+    aspect.ratio = 1
+  ) +
+  guides(color = guide_legend(override.aes = list(alpha = 1, size = 1.5)))
```
I sort of wonder if we should just turn off the legend actually, since it's not meaningful across panels?

sjspielman commented on analyses/cell-type-ewings/template_notebooks/auc-singler-workflow/03-clustering.Rmd:
```diff
@@ -0,0 +1,265 @@
+Below we calculate a series of statistics: 
+
+- Average silhouette width: This metric evaluates cluster separation. 
+Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters. 
```
I'd add a "higher = ..." sentence here. Not convinced I have the right wording here, but along these lines.

```suggestion
Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters. 
Higher values indicate tighter clusters.
```

sjspielman commented on analyses/cell-type-ewings/template_notebooks/auc-singler-workflow/03-clustering.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+- Average silhouette width: This metric evaluates cluster separation. 
+Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters. 
+- Average cluster purity: This metric also evaluates cluster separation and tells us the proportion of neighboring cells that are assigned to the same cluster. 
+Higher purity values indicate that clusters are well separated. 
+- Cluster stability: This evaluates how stable the clustering is to input data. 
+Higher values of cluster stability indicate more reproducible clusters. 
```
I would say plainly that these range between 0-1 (which is different from silhouette)

sjspielman commented on analyses/cell-type-ewings/template_notebooks/auc-singler-workflow/03-clustering.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+## Marker gene expression across clusters 
+
+Finally, we will look at the expression of marker genes across each cluster for each value of K. 
```
```suggestion
Finally, we will look at the expression of marker genes across each cluster for each value of `k`. 
```

sjspielman commented on analyses/cell-type-ewings/template_notebooks/auc-singler-workflow/03-clustering.Rmd:
````diff
@@ -0,0 +1,265 @@
+
+```{r}
+# read in marker genes table
+marker_genes_df <- readr::read_tsv(params$marker_genes_file, show_col_types = FALSE) |>
````
I don't think you need this since you set the option at the top

```suggestion
marker_genes_df <- readr::read_tsv(params$marker_genes_file) |>
```

sjspielman commented on analyses/cell-type-ewings/template_notebooks/auc-singler-workflow/03-clustering.Rmd:
```diff
@@ -0,0 +1,265 @@
+
+k_params |>
+  purrr::map(\(k) {
+    plot_marker_genes(cluster_exp_df,
```
Looking at this again in this template, I think we should wrap these plots, and/or maybe also only look at not-uncommon cell types (maybe the top 10 or so, and have a 5x2 grid?). Otherwise this section could get really long and out of hand fast!

allyhawkins replied:
Just an FYI that this will only ever look at tumor, immune, endothelial, and mesenchymal like cells. So there will always just be 4 density plots per value of k.

sjspielman reviewed (commented):
This mostly looks good to me! I left a couple organizational comments as well as general plotting suggestions.

sjspielman commented:
> I also included a section where we compare the clusters to the SingleR annotations obtained from aucell-singler-annotation.sh. Because this notebook will be somewhat dependent on running that workflow I actually think I might add rendering this as the last step in that workflow rather than creating a whole separate workflow? I didn't do that in this PR, but that is why I am adding this as 03-clustering.Rmd to the template_notebooks/aucell-singler-workflow folder. Let me know if you think I should move this and ultimately create a separate workflow.

I'm of two minds here, but ultimately I think I prefer that this be in its own "evaluate clustering" workflow to run after the singler one?

allyhawkins commented:
> I'm of two minds here, but ultimately I think I prefer that this be in its own "evaluate clustering" workflow to run after the singler one?

I see both ways and as I think about it we may want the ability to expand this even more so I agree about making it separate. I moved this template into it's own folder to reflect that ultimately it will go into its own workflow. 

I addressed the other comments you had or replied them to inline including removing the legend from the UMAPs. I agree that they are kind of meaningless in terms of the actual numbers. 

The main change I made was to the last set of plots. I made it so for each value of k the density plots are 2x2. I also updated the y-axis title to be "Cluster" since it was set by default as "Cell type annotation" in the original function. 

Let me know if I missed anything! Here's a copy of the updated report: 
[01-clustering.html.zip](https://github.com/user-attachments/files/16512845/01-clustering.html.zip)

sjspielman reviewed (approved):
> The main change I made was to the last set of plots. I made it so for each value of k the density plots are 2x2. I also updated the y-axis title to be "Cluster" since it was set by default as "Cell type annotation" in the original function.

These look much better!

Everything else looks good to me, let's go! 🚀
